_posts/2022-11-22-redis-job-queue-reloaded.adoc

```diff
@@ -0,0 +1,183 @@
+
+image::/assets/images/posts/redis-job-queue/reloaded.png[align="center"]
+
+These private queues form a safety net.
+
+So, the simulator does use not only the _main queue_ but also its private queue:
+
+[source, java]
+----
+this.queues = ds.list(FightRequest.class);
+this.queueName = "queue-" + name; // the name of the private queue
+----
+
+When a simulator pops a request from the main queue, it does not process it immediately; it writes it to its private queue.
+To achieve this, we cannot use `brpop` and then writes to the other queue, as if something wrong happens in between, we would have the same problem.
+Instead, we use `blmove,` which pops an element from a list and pushes it into another in an atomic fashion. 
+Thus, we ensure that multiple simulators cannot consume the same request and that the request cannot be lost.
+
+So, we use the following code to move the request from the main queue to the private queue:
+
+[source, java]
+----
+// pop the item at the right side of the 'fight-requests' queue
+// and writes it to the left side of 'queueName'.
+// it returns the moved item or `null` in the entry queue, 'fight-requests',
+// does not have any item, even after the 1-second delay
+var moved = queues.blmove("fight-requests", queueName,  
+        Position.RIGHT, Position.LEFT, Duration.ofSeconds(1));
+----
+
+Now, the simulator does not simulate the requests from the main queue but needs to process the ones added to the private queue.
+
+[source, java]
+----
+public void processRequestFromPrivateQueue() {  
+    var request = queues.lindex(queueName, -1);  
+    while (request != null) {  
+        runSimulation(request);  
+        queues.lrem(queueName, 1, request);    
+        request = queues.lindex("queue-" + name, -1);  
+    }  
+}
+----
+
+This code is slightly different from the code from the previous blog.
+This time, we do not pop. 
+We get the last item from the queue (index `-1` is the last one), process it, and then remove it from the queue.
+We do this until the queue is empty.
+
+Let's put everything together:
+1. when the simulator starts, it should process the items from its private queue. So, if it crashes, it will retry to process the item.
+2. once the private queue is empty, it gets new requests from the main queue. It will not process it directly but re-trigger the processing of the private queue until the queue is empty.
+
+[source, java]
+----
+@Override  
+public void run() {  
+    // First, check if we are recovering, and drain the requests from the 
+    // simulator's queue  
+  processRequestFromPrivateQueue();  
+  while (! stopped) {  
+    // Simulator's queue drained - poll the main queue  
+    var moved = queues.blmove("fight-requests", queueName,  
+            Position.RIGHT, Position.LEFT, Duration.ofSeconds(1));  
+    if (moved != null) {  
+        // If an element has been moved, process it  
+        processRequestFromPrivateQueue();  
+    }  
+  }
+}
```

Comment:
The indentation is a bit off in this excerpt. The comments at the top are indented too much. And it looks like it mixes 2 spaces and 4 spaces indent.

Reply:
I added the comments in my draft (on obsidian), and something weird happened...